### PR TITLE
build: bump to use go1.21

### LINKED ISSUE #284: Build with go1.21
Build with the latest and greatest go1.21.0

### PR DESCRIPTION
**Pull request by chenrui333**
- build: bump to use go1.21  

Resolves #284

### CODE REVIEW

**Comment by cube2222** [comment]
Hello @chenrui333, thanks for the contribution!

Please read the [contributing guide](https://github.com/opentffoundation/opentf/blob/main/CONTRIBUTING.md) and create an issue before contributing, I'll be moving this PR to draft status.

**Comment by chenrui333** [comment]
@cube2222 filed an issue https://github.com/opentffoundation/opentf/issues/284

**Review by cube2222 — DISMISSED** [review]
LGTM!

Thanks for the contribution once again @chenrui333!

We'll merge once we get one more approval.

**Comment by cube2222** [comment]
Seems like a test is broken @chenrui333, could you please take a look?

**Comment by Yantrio** [comment]
Hi @chenrui333 it seems that this change has broken a unit test: https://github.com/opentffoundation/opentf/actions/runs/6090005672/job/16524303458?pr=282#step:6:152

As far as I remember golang 1.21 did some improvements to TLS verification that may have caused this.

**Comment by chenrui333** [comment]
yeah, digging the test :)

**Comment by Yantrio** [comment]
I had a vague recollection of this effecting other projects I was working on recently and to help you fix the test I routed the change out from the golang changes here: https://tip.golang.org/doc/go1.21

> The TLS alert codes sent from the server for client authentication failures have been improved. Previously, these failures always resulted in a "bad certificate" alert. Now, certain failures will result in more appropriate alert codes, as defined by RFC 5246 and RFC 8446:

> - For TLS 1.3 connections, if the server is configured to require client authentication using [RequireAnyClientCert](https://tip.golang.org/pkg/crypto/tls/#RequireAnyClientCert) or [RequireAndVerifyClientCert](https://tip.golang.org/pkg/crypto/tls/#RequireAndVerifyClientCert), and the client does not provide any certificate, the server will now return the "certificate required" alert.

**Comment by alrs** [comment]
Are you sure you want to do this? The version of Go packaged in Debian stable is only 1.19. Go developers will obviously have the latest release on their systems, many end-users will not.

https://packages.debian.org/bookworm/golang

**Comment by Yantrio** [comment]
Just a note for future reference, I personally would like to ensure that we go through the rest of the golang 1.21 changes and make sure that they dont effect opentf in any negative way.

Whilst this small issue has happened around a broken unit test for mTLS, we should investigate the other changes in go 1.21 just incase.

**Comment by chenrui333** [comment]
just some test update, should be good now.

**Comment by chenrui333** [comment]
> Are you sure you want to do this? The version of Go packaged in Debian stable is only 1.19. Go developers will obviously have the latest release on their systems, many end-users will not.
> 
> https://packages.debian.org/bookworm/golang

just FYI, go1.19 is EOL already when go1.21 came out.

**Comment by chenrui333** [comment]
If you folks have concerns about the go compatibility, please read this post, https://go.dev/blog/compat

Also I am a homebrew maintainer myself, and we are pretty eager to push all folks to use the latest go (not saying that we would like to disrupt everyone's flows, but just saying I dont see much issues from my perspective)

**Comment by cube2222** [comment]
@chenrui333 I believe the test should check for the part @Yantrio highlighted, so "certificate required". Right now it's just testing an arbitrary error getting the state.

But I also agree with @Yantrio that we should go through the list of changes to double-check if there isn't anything else possibly affecting us.

**Comment by kamikazechaser** [comment]
> Go developers will obviously have the latest release on their systems, many end-users will not.

This change will only affect developers using the Debian golang package (Which btw is not the official recommended way of setting up Go on Linux). End users will still consume opentf binaries.

**Comment by kamikazechaser** [comment]
> Go developers will obviously have the latest release on their systems, many end-users will not.

This change will only affect developers using the Debian golang package (Which btw is not the official recommended way of setting up Go on Linux). End users will still consume opentf binaries.

**Comment by kamikazechaser** [comment]
https://go.dev/blog/toolchain

This change is fine because of the lack of `pkg`s that a 3rd party can rely on. But it does have implications if there is a future decision around this.

File: internal/backend/remote-state/http/server_test.go
Comment (eranelbaz) [inline internal/backend/remote-state/http/server_test.go:288]: why this change is relevant?

Comment (eranelbaz) [reply internal/backend/remote-state/http/server_test.go:288]: https://github.com/opentffoundation/opentf/pull/282#issuecomment-1707376275
https://github.com/opentffoundation/opentf/pull/282/commits/1fe22d23c8b465bc0740ffadbeae3ff3668f2dd9
Ok resolving

**Review by cube2222 — CHANGES_REQUESTED** [review]
Just so it doesn't get drowned out in the other comments, see https://github.com/opentffoundation/opentf/pull/282#issuecomment-1707388794

**Comment by Yantrio** [comment]
I've added a comment to #282 to reflect on what we need to look into before upgrading to 1.21.

In an ideal world, this version bump would be a simple process, but sadly life is never that simple :(

**Comment by chenrui333** [comment]
I dont have enough bandwidth for digging the full list between go1.20 vs go1.21 (especially regarding mTLS), so help is needed.

**Comment by cube2222** [comment]
@chenrui333 Makes sense. We'll get to this eventually, but it's not a priority right now.

**Comment by dvaumoron** [comment]
https://github.com/chenrui333/opentf/blob/go-1.21/internal/backend/remote-state/http/server_test.go#L287-L288

seems a strange correction, i think a non tls error should still be reported as failure, so "remote error: tls:" instead of "remote error: tls: bad certificate" would be a better change

**Comment by Yantrio** [comment]
Closing in favor of #718.

**Pull request closed without merging**
